src/ffi.rs

```diff
@@ -260,6 +260,49 @@ extern "C" {
 }
 
 
+#[cfg(feature = "std")]
+#[no_mangle]
+/// A reimplementation of the C function `secp256k1_context_create` in rust.
+///
+/// This function allocates memory, the pointer should be deallocated using `secp256k1_context_destroy`
+/// A failure to do so will result in a memory leak.
+///
+/// This will create a secp256k1 raw context.
+// Returns: a newly created context object.
+//  In:      flags: which parts of the context to initialize.
+pub unsafe extern "C" fn secp256k1_context_create(flags: c_uint) -> *mut Context {
+    assert!(mem::align_of::<usize>() >= mem::align_of::<u8>());
+
+    let size: usize =  secp256k1_context_preallocated_size(flags)/mem::size_of::<usize>();
+
+    let buf = vec![0usize; size + 1].into_boxed_slice();
+    let ptr = Box::into_raw(buf) as *mut usize;
+    ::core::ptr::copy_nonoverlapping(&size, ptr, 1);
+    let ptr: *mut usize = ptr.offset(1);
+
+    secp256k1_context_preallocated_create(ptr as *mut c_void, flags)
+}
+
+#[cfg(feature = "std")]
+#[no_mangle]
+/// A reimplementation of the C function `secp256k1_context_destroy` in rust.
+///
+/// This function destroys and deallcates the context created by `secp256k1_context_create`.
+///
+/// The pointer shouldn't be used after passing to this function, consider it as passing it to `free()`.
+///
+pub unsafe extern "C" fn secp256k1_context_destroy(ctx: *mut Context) {
+    assert_eq!(ctx as usize % mem::align_of::<usize>(), 0);
```

Comment:
I think this assertion should be dropped; it's not obvious to me that alignment should be preserved across casts to `usize` like this and ultimately if the user gives us a pointer here that we didn't create then it's UB anyway.

Reply:
shouldn't it hopefully be the same pointer that was created by `secp256k1_context_create`?
I think of this assertion as a fail safe to notice us if something went really wrong on some weird system. but I guess it can be removed

Reply:
There exist systems where the low-order bits of pointers are used for weird accounting stuff, so the alignment of the pointer will not correspond to the "alignment" of the result of casting to `usize`.

Reply:
Weird. people here suggested this to be the right way to check alignment https://github.com/rust-lang/rust/issues/42561

Reply:
Removed anyway.